src/ActionMenu.tsx

```diff
@@ -21,11 +23,26 @@ const ActionMenuBase = ({
   renderAnchor = <T extends ButtonProps>(props: T) => <Button {...props} />,
   renderItem = Item,
   onAction,
+  open,
+  setOpen,
   ...listProps
 }: ActionMenuProps): JSX.Element => {
-  const [open, setOpen] = useState(false)
-  const onOpen = useCallback(() => setOpen(true), [])
-  const onClose = useCallback(() => setOpen(false), [])
+  const [internalOpen, setInternalOpen] = useState(false)
+  const openState = open ?? internalOpen
+
+  const onOpen = useCallback(() => {
+    setInternalOpen(true)
+    if (setOpen) {
+      setOpen(true)
+    }
+  }, [setOpen])
+
+  const onClose = useCallback(() => {
+    setInternalOpen(false)
+    if (setOpen) {
+      setOpen(false)
+    }
+  }, [setOpen])
```

Comment:
This seems like a pattern that we will want to replicate into DropdownMenu and SelectPanel.  What are your thoughts on pulling out a hook that could be used like `[open, setOpen] = useOptionallyExternalState(externalOpen, setExternalOpen, false)` (definitely open to different naming).  Then we would only have one state and not maintain both internal and external.

Reply:
@dgreif yeah, I think that would clean things up quite a bit! To be clear, I think we would still want to be maintaining the internal state within the `useOptionallyExternalState` hook since we could not conditionally implement hooks - we just wouldn't expose that conditional internal state outside of the hook. Is that basically what you are suggesting?

Reply:
@VanAnderson something similar to https://github.com/primer/components/blob/main/src/hooks/useProvidedRefOrCreate.ts.  It would create the internal state, but not use it if an external state was provided.  `useProvidedStateOrCreate` may be more in-line now that I see this other name 😄